chips/rp2040/src/uart.rs

```diff
@@ -0,0 +1,733 @@
+        self.disable();
+        self.registers.uartlcr_h.modify(UARTLCR_H::FEN::CLEAR);
+
+        let clk = 125_000_000;
+
+        // Calculate baud rate
+        let baud_rate_div = 8 * clk / params.baud_rate;
+        let mut baud_ibrd = baud_rate_div >> 7;
+        let mut baud_fbrd = ((baud_rate_div & 0x7f) + 1) / 2;
+
+        if baud_ibrd == 0 {
+            baud_ibrd = 1;
+            baud_fbrd = 0;
+        } else if baud_ibrd >= 65535 {
+            baud_ibrd = 65535;
+            baud_fbrd = 0;
+        }
+
+        self.registers
+            .uartibrd
+            .write(UARTIBRD::BAUD_DIVINT.val(baud_ibrd));
+        self.registers
+            .uartfbrd
+            .write(UARTFBRD::BAUD_DIVFRAC.val(baud_fbrd));
+
+        self.registers.uartlcr_h.modify(UARTLCR_H::BRK::SET);
+        // Configure the word length
+        match params.width {
+            Width::Six => self.registers.uartlcr_h.modify(UARTLCR_H::WLEN::BITS_6),
+            Width::Seven => self.registers.uartlcr_h.modify(UARTLCR_H::WLEN::BITS_7),
+            Width::Eight => self.registers.uartlcr_h.modify(UARTLCR_H::WLEN::BITS_8),
+        }
+
+        // Configure parity
+        match params.parity {
+            Parity::None => {
+                self.registers.uartlcr_h.modify(UARTLCR_H::PEN::CLEAR);
+                self.registers.uartlcr_h.modify(UARTLCR_H::EPS::CLEAR);
+            }
+
+            Parity::Odd => {
+                self.registers.uartlcr_h.modify(UARTLCR_H::PEN::SET);
+                self.registers.uartlcr_h.modify(UARTLCR_H::EPS::CLEAR);
+            }
+            Parity::Even => {
+                self.registers.uartlcr_h.modify(UARTLCR_H::PEN::SET);
+                self.registers.uartlcr_h.modify(UARTLCR_H::EPS::SET);
+            }
+        }
+
+        // Set the stop bit length - 2 stop bits
+        match params.stop_bits {
+            StopBits::One => self.registers.uartlcr_h.modify(UARTLCR_H::STP2::CLEAR),
+            StopBits::Two => self.registers.uartlcr_h.modify(UARTLCR_H::STP2::SET),
+        }
+
+        // Set flow control
+        if params.hw_flow_control {
+            self.registers.uartcr.modify(UARTCR::RTSEN::SET);
+            self.registers.uartcr.modify(UARTCR::CTSEN::SET);
+        } else {
+            self.registers.uartcr.modify(UARTCR::RTSEN::CLEAR);
+            self.registers.uartcr.modify(UARTCR::CTSEN::CLEAR);
+        }
+        self.registers.uartlcr_h.modify(UARTLCR_H::BRK::CLEAR);
+
+        // FIFO is not precise enough for receive
+        self.registers.uartlcr_h.modify(UARTLCR_H::FEN::CLEAR);
+
+        // Enable uart and transmit
+        self.registers
+            .uartcr
+            .modify(UARTCR::UARTEN::SET + UARTCR::TXE::SET + UARTCR::RXE::SET);
+
+        self.registers
+            .uartdmacr
+            .write(UARTDMACR::TXDMAE::SET + UARTDMACR::RXDMAE::SET);
+
+        let n = unsafe { cortexm0p::nvic::Nvic::new(self.interrupt) };
+        n.enable();
```

Comment:
flagging this as also hopefully not needed

Reply:
done